protocols/rendezvous/src/substream_handler.rs

```diff
@@ -0,0 +1,463 @@
+// Copyright 2021 COMIT Network.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A generic [`ProtocolsHandler`] that delegates the handling of substreams to [`SubstreamHandler`]s.
+//!
+//! This module is an attempt to simplify the implementation of protocols by freeing implementations from dealing with aspects such as concurrent substreams.
+//! Particularly for outbound substreams, it greatly simplifies the definition of protocols through the [`FutureSubstream`] helper.
+//!
+//! At the moment, this module is an implementation detail of the rendezvous protocol but the intent is for it to be provided as a generic module that is accessible to other protocols as well.
+
+use futures::future::{BoxFuture, Fuse, FusedFuture};
+use futures::FutureExt;
+use libp2p_core::upgrade::FromFnUpgrade;
+use libp2p_core::Endpoint;
+use libp2p_swarm::protocols_handler::{InboundUpgradeSend, OutboundUpgradeSend};
+use libp2p_swarm::{
+    KeepAlive, NegotiatedSubstream, ProtocolsHandler, ProtocolsHandlerEvent,
+    ProtocolsHandlerUpgrErr, SubstreamProtocol,
+};
+use std::collections::{HashMap, VecDeque};
+use std::fmt;
+use std::future::{Future, Ready};
+use std::hash::Hash;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+use void::Void;
+
+/// Handles a substream throughout its lifetime.
+pub trait SubstreamHandler: Sized {
+    type InEvent;
+    type OutEvent;
+    type Error;
+    type OpenInfo;
+
+    fn new(substream: NegotiatedSubstream, info: Self::OpenInfo) -> Self;
+    fn inject_event(self, event: Self::InEvent) -> Self;
+    fn advance(self, cx: &mut Context<'_>) -> Result<Next<Self, Self::OutEvent>, Self::Error>;
```

Comment:
The core aspect of this design is actually that it takes `self` by value instead of as a reference.

Looking at implementations of substream state machines, it is common that you actually need ownership to advance the state machine which results in `std::mem::replace` being used all the time. By providing an interface that gives you `self` directly instead of `&mut self`, I hoped to centralize the use of `std::mem::replace` into one place (`SubstreamProtocolsHandler`) and at the same time, it allows for a neat way of "shutting" down a `SubstreamHandler` why dropping the internal stream.

Reply:
From your review post:

> Have you considered a SubstreamHandler to implement Sink instead of inject_event and Stream instead of advance?

Above mentioned reason is also why I think this is not going to work as well because all `Sink`/`Stream` methods take `&mut self`.